core/trino-main/src/main/java/io/trino/operator/DriverFactory.java

```diff
@@ -96,12 +97,37 @@ public synchronized Driver createDriver(DriverContext driverContext)
     {
         checkState(!noMoreDrivers, "noMoreDrivers is already set");
         requireNonNull(driverContext, "driverContext is null");
-        ImmutableList.Builder<Operator> operators = ImmutableList.builder();
-        for (OperatorFactory operatorFactory : operatorFactories) {
-            Operator operator = operatorFactory.createOperator(driverContext);
-            operators.add(operator);
+        List<Operator> operators = new ArrayList<>();
+        try {
+            for (OperatorFactory operatorFactory : operatorFactories) {
+                Operator operator = operatorFactory.createOperator(driverContext);
+                operators.add(operator);
+            }
+            return Driver.createDriver(driverContext, operators);
+        }
+        catch (Throwable failure) {
+            for (Operator operator : operators) {
+                try {
+                    operator.close();
+                }
+                catch (Throwable closeFailure) {
+                    if (failure != closeFailure) {
+                        failure.addSuppressed(closeFailure);
+                    }
+                }
+            }
+            for (OperatorContext operatorContext : driverContext.getOperatorContexts()) {
```

Comment:
Do we know that at this point the operator context which are present in driverContext only correspond to operators created above?

Reply:
It is possible that Operator creation fails after the corresponding `OperatorContext` has already been created

Reply:
Yeah - that I understand. My quesiton is can there be context from other operators, which should not be closed here?

Reply:
There shouldn't be any other OperatorContext's but the ones for a driver that we failed to created

Reply:
Yeah - looked into the code. I though there should be 1:1 mapping between Driver and DriverContext, just it looked weird that both are not created at the same time. I guess some optimization.